core/src/main/java/org/apache/calcite/rex/RexUtil.java

```diff
@@ -514,9 +514,13 @@ public Boolean visitDynamicParam(RexDynamicParam dynamicParam) {
     }
 
     public Boolean visitCall(RexCall call) {
-      // Constant if operator is deterministic and all operands are
-      // constant.
-      return call.getOperator().isDeterministic()
```

Comment:
Based on this comment of `isDeterministic`, I am thinking the original idea of using `isDeterministic()` is still correct.

Basically `current_timestamp` is not deterministic depends on the time to call it.

Instead of changing here, I found that `current_timestamp`, `current_date`, etc. return `true` for  `isDeterministic()`. Should those functions be changed to return `fasle`?

Reply:
`current_timestamp` is deterministic, or non-volatile. Because the value doesn't change within a transaction. Functions like random(), timeofday() are volatile, because the value can change in a single query, we need to reevaluate for every tuple.

Reply:
Danny mentions in the JIRA that if it's a batch job that runs cross days, the result is wrong. 

My understanding is if `current_timestamp` is called in different day because batch job could be large, then it returns a different `current_timestamp`,  do I misunderstand it? If my understanding is correct, then if we say a batch query is in a transaction, it sounds more like `current_timestamp` implementation is wrong.

I found that in SqlFunctions, `current_timestamp` is marked as non-deterministic: https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java#L2235

Reply:
It depends on different database vendors. MySQL, Postgres mark it as deterministic, while SQL Server mark it as nondeterministic.

https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_current-timestamp
https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_now
NOW() returns a constant time that indicates the time at which the statement began to execute.

Reply:
And Vertica:
https://www.vertica.com/docs/9.2.x/HTML/Content/Authoring/SQLReferenceManual/Functions/Date-Time/CURRENT_TIMESTAMP.htm

Reply:
Depends on the definition of "deterministic":
* returns the same value every time you call it? (not true for CURRENT_TIMESTAMP)
* returns the same value every time you call it within a given query execution (true for CURRENT_TIMESTAMP)
* returns the same value every you call it within a given row (also true for CURRENT_TIMESTAMP, also true for the RANDOM function in some databases IIRC)

I don't recall how Calcite defines "deterministic". Different optimizations need different notions of "constant" and "determinism".